mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp

```diff
@@ -55,6 +56,24 @@ static void strToTokens(const std::string &arguments,
     }
   }
 }
+
+bool lazy_init() {
+  static const bool once = []() {
```

Comment:
[Recommendation] Initialization of static is thread-safe since C++11, see https://en.cppreference.com/w/cpp/language/storage_duration. Therefore you do not need to protect lazy_init with mutex.

Reply:
@atamazov Yah, I'm aware of that. Though if I move `lazy_init()` out of the critical section there is still a chance that:
 - thread A is executing lowering (in mutex)
 - thread B is executing initialization (out of mutex)
of what we originally worried about.

Considering that initialization only run once, that scenario is probably quite rare. This is more of a better safe than sorry approach.

Reply:
Initialization of static objects happens only once. If A is executing lowering, then initialization is already done. B will skip initialization and wait on mutex.

Reply:
But you can keep this as is

Reply:
@atamazov Thanks for reminding the static initialization happens only once. Now that I think about it, I don't think mutex is needed at all: 
  - For initialization, we don't need to worry about race condition because it is atomic
  - This way, no concerns in multi-threaded lowering too because multi-threaded lowering happens in independent pass managers.

Reply:
Good!